gossipsub.go

```diff
@@ -1111,6 +1173,15 @@ func (gs *GossipSubRouter) emitGossip(topic string, exclude map[peer.ID]struct{}
 		return
 	}
 
+	// shuffle to emit in random order
+	shuffleStrings(mids)
+
+	// if we are emitting more than GossipSubMaxIHaveLength mids, truncate the list
+	if len(mids) > GossipSubMaxIHaveLength {
+		log.Debugf("too many messages for gossip; truncating IHAVE list (%d messages)", len(mids))
+		mids = mids[:GossipSubMaxIHaveLength]
```

Comment:
Would it make more sense to shuffle by peer? Or maybe shuffle once, then create as many partitions as peers we are gossiping to? I'm thinking we want to get the word out there to _someone_ for as many messages as we can, instead of picking a set and cloning the gossip for all peers?

Reply:
This is a good point, and we have enough redundancy in the system that even if we truncate the list we'll get multiple IHAVEs and eventually all the messages.
Only downside is that we need to copy the slice.

Reply:
done.